### ISSUE TITLE
why is "Array of plugins" deprecated

### ISSUE THREAD

**Issue by jorisgu** [bangle-io/bangle.dev#175]
TL;DR:

When building the array of plugins (`myPlugins`)  I have to pass it through extra code `()=>myPlugins` to remove a warning. 

________________________________________________________

Passing an array of plugins to `BangleEditorState` is throwing a warning:
```
    if (Array.isArray(plugins)) {
      console.warn(
        'The use plugins as an array is deprecated, please pass a function which returns an array of plugins. Refer: https://bangle.dev/docs/api/core#bangleeditorstate',
      );
    }
```
But then this array go through `flatten` and `recurse` methods which seems to completely handle arrays. 

Could you please explain why this is deprecated ? 
Do you want to enforce using function for some specific pattern to come ? 

The documentation at [https://bangle.dev/docs/api/core#bangleeditorstate](https://bangle.dev/docs/api/core#bangleeditorstate) presents the parameter `plugins` as a `plugins: ?({ schema, specRegistry }) -> Plugins[]`. Maybe this is to enforce to always build plugin with schema or specRegisty available as present in `flatten`?

```
    if (typeof plugins === 'function') {
      if (!callbackPayload) {
        throw new Error('Found a function but no payload');
      }
      return recurse(plugins(callbackPayload));
    }
```

Would you recommend to always mimic the plugin factory like in every classic component?
For reference the italic plugin:
```
function pluginsFactory({ keybindings = defaultKeys } = {}) {
  return ({ schema }) => {
    const type = getTypeFromSchema(schema);

    return [
      markPasteRule(/_([^_]+)_/g, type),
      markPasteRule(/\*([^*]+)\*/g, type),
      markInputRule(/(?:^|[^_])(_([^_]+)_)$/, type),
      markInputRule(/(?:^|[^*])(\*([^*]+)\*)$/, type),
      keybindings &&
        keymap({
          [keybindings.toggleItalic]: toggleMark(type),
        }),
    ];
  };
}
```

**Comment by kepta** [comment]
> Could you please explain why this is deprecated ?
Do you want to enforce using function for some specific pattern to come ?

Array has been deprecated for two reasons:
- Lazy evaluation of plugins.
- Have parity with React api.

Earlier I was allowing for directly providing an array of plugins and this lead to problems as every time react element will be Re-rendered the entire plugins array would be recreated. This is specifically problematic for plugins which create an expensive state. 

> Would you recommend to always mimic the plugin factory like in every classic component?

I would recommend that for complicated plugins that need access to `schema`. Other than that it should be fine to just have a thin function wrapper like in this [placeholder example](https://bangle.dev/docs/examples/placeholder).

**Comment by jorisgu** [comment]
Ok,
Thank you for those precisions!
